include/deal.II/base/symmetric_tensor.h

```diff
@@ -3521,7 +3521,216 @@ outer_product(const SymmetricTensor<2, dim, Number> &t1,
   return tmp;
 }
 
+/**
+ * Perform a spectrum decomposition of a 2nd-order symmetric tensor \a
+ * original_tensor given as the input argument, \f[ \mathrm{original\_tensor} =
+ * \sum_i \lambda_i \, \boldsymbol{n}_i \otimes \boldsymbol{n}_i, \f] where
+ * $\lambda_i$ is the eigenvalue, and $\boldsymbol{n}_i$ is the corresponding
+ * eigenvector. The outputs are the positive part \a positive_part_tensor and
+ * negative part \a negative_part_tensor of the input tensor,
+ * that is,
+ * \f[
+ *   \mathrm{positive\_part\_tensor} = \sum_i <\lambda_i>_+ \boldsymbol{n}_i
+ * \otimes \boldsymbol{n}_i, \quad \mathrm{negative\_part\_tensor} = \sum_i
+ * <\lambda_i>_- \boldsymbol{n}_i \otimes \boldsymbol{n}_i, \f] where
+ * $<\lambda_i>_+ = \mathrm{max}\{ \lambda_i, 0 \}$ and
+ * $<\lambda_i>_- = \mathrm{min}\{ \lambda_i, 0 \}$. Obviously,
+ * \f[
+ *   \mathrm{positive\_part\_tensor}  + \mathrm{negative\_part\_tensor} =
+ * \mathrm{original\_tensor}. \f]
+ *
+ * @param[in] original_tensor The 2nd-order symmetric tensor to be split into
+ * the positive and negative parts
+ * @param[out] positive_part_tensor The positive part of the input tensor in
+ * which the eigenvalues are positive or zero
+ * @param[out] negative_part_tensor The negative part of the input tensor in
+ * which the eigenvalues are negative or zero
+ *
+ * @relatesalso SymmetricTensor
+ */
+template <int dim, typename Number>
+void
+positive_negative_split(const SymmetricTensor<2, dim, Number> &original_tensor,
+                        SymmetricTensor<2, dim, Number> &positive_part_tensor,
+                        SymmetricTensor<2, dim, Number> &negative_part_tensor)
```

Comment:
My preference would be if this function returned the two parts of the tensor as a std::pair<SymmetricTensor,SymmetricTensor>`. In C++17, you can then used [structured bindings](https://en.cppreference.com/w/cpp/language/structured_binding) to declare variables from the result, and also make them `const`.

Reply:
@bangerth Hi Wolfgang, I have not used the structured bindings before, so please let me know what I did was correct:
```
template <int dim, typename Number>
std::pair<SymmetricTensor<2, dim, Number>, SymmetricTensor<2, dim, Number>>
positive_negative_split(const SymmetricTensor<2, dim, Number> &original_tensor)
{
  Assert(dim <= 3, ExcNotImplemented());

  const std::array<std::pair<Number, Tensor<1, dim, Number>>, dim>
    eigen_system = eigenvectors(original_tensor);

  std::pair<SymmetricTensor<2, dim, Number>, SymmetricTensor<2, dim, Number>>
    postive_negative_tensors;

  auto & [positive_part_tensor, negative_part_tensor] = postive_negative_tensors;

  positive_part_tensor = 0;
  for (unsigned int i = 0; i < dim; ++i)
    if (eigen_system[i].first > 0)
      positive_part_tensor += eigen_system[i].first *
                              symmetrize(outer_product(eigen_system[i].second,
                                                       eigen_system[i].second));

  negative_part_tensor = 0;
  for (unsigned int i = 0; i < dim; ++i)
    if (eigen_system[i].first < 0)
      negative_part_tensor += eigen_system[i].first *
                              symmetrize(outer_product(eigen_system[i].second,
                                                       eigen_system[i].second));
  return postive_negative_tensors;
}
```
When I call the above function, 
`  const auto & [positive_tensor, negative_tensor] = positive_negative_split(symmetric_tensor);
`

Please let me know whether this is what you meant.